### ISSUE TITLE
false positive "Possible null pointer dereference due to return value of called method"

### ISSUE THREAD

**Issue by florianrosenauer** [spotbugs/sonar-findbugs#255]
## Issue Description
We noticed some findings from sonar-findbugs where in my opinion no NPE is possible but it is flagged that way.

It raises the "Possible null pointer dereference in App.getFileName(Path) due to return value of called method" = findbugs:NP_NULL_ON_SOME_PATH_FROM_RETURN_VALUE althought it is checked not to be null before.

Screen from the analysis:
![reproducer-screen-edit](https://user-images.githubusercontent.com/5151028/53407313-4671aa80-39bc-11e9-8dff-8bf49b5d5f50.png)

## Environment

| Component          | Version |
| ------------------ | ------- |
| SonarQube          | 6.7.2 |
| Sonar-FindBugs     | 3.9.2 |
| Maven              | 3.5.4 |
| Gradle             | na |
| Java               | Oracle 1.8.181 |

## Code (If needed)
If this code is analyzed, it raises the NPE rule at the last return statement:

```java
package sonarqube.test.sonarqube_npetest;

import java.nio.file.Path;

/**
 * Hello world!
 *
 */
public class App 
{
    public static void main( String[] args )
    {
        System.out.println( "Hello Java!" );
    }
    
	protected String getFileName(Path path) {
		if (path == null || path.getFileName() == null) {
			System.out.println( "Hello Java!" );
			return "Hello Java!";
		}
		
		System.out.println( "Hello Java!" );
		return path.getFileName().toString();
	}
	
}

```

**Comment by KengoTODA** [comment]
We human expect that the first `getFileName()` invocation and second one return the same result, but machine does not. Actually no logic supports our expectation. To make it sure, store returned value to local variable, then computer agrees that the filename isn't null at the last `return`.

```java
	protected String getFileName(Path path) {
		if (path == null) {
			return "Hello Java!";
                } 
                String fileName = path.getFileName();
                if (fileName == null) {
			return "Hello Java!";
		} else {
		        return path.getFileName().toString();
                }
	}
```

In other words, this is not false-positive in the following case, and nobody can prove that it's not true for `Path` class (technically possible, though).

```java
int count = 0;
Path getFileName() {
    count++;
    if (count == 1) {
        return fileName;
    } else {
        return null;
    }
}
```
